log_test.go

```diff
@@ -0,0 +1,108 @@
+// +build !integration,unit
+
+package wavelet
+
+import (
+	"encoding/hex"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/perlin-network/noise/skademlia"
+	"github.com/perlin-network/wavelet/log"
+	"github.com/perlin-network/wavelet/sys"
+	"github.com/stretchr/testify/assert"
+	"github.com/valyala/fastjson"
+)
+
+func TestCollapseResultsLogger(t *testing.T) {
+	log.ClearWriters()
+
+	logger := NewCollapseResultsLogger()
+
+	keys, err := skademlia.NewKeys(1, 1)
+	if !assert.NoError(t, err) {
+		return
+	}
+
+	recipient, err := skademlia.NewKeys(1, 1)
+	if !assert.NoError(t, err) {
+		return
+	}
+
+	results := &collapseResults{}
+
+	payload, err := Transfer{Recipient: recipient.PublicKey(), Amount: 1}.Marshal()
+	if !assert.NoError(t, err) {
+		return
+	}
+	txApplied := NewTransaction(keys, 0, 0, sys.TagTransfer, payload)
+	results.appliedCount++
+	results.applied = append(results.applied, &txApplied)
+
+	payload, err = Transfer{Recipient: recipient.PublicKey(), Amount: 10}.Marshal()
+	if !assert.NoError(t, err) {
+		return
+	}
+	txRejected := NewTransaction(keys, 0, 0, sys.TagTransfer, payload)
+	results.rejectedCount++
+	results.rejected = append(results.rejected, &txRejected)
+	results.rejectedErrors = append(results.rejectedErrors, errors.New("error rejected"))
+
+	logCh := make(chan []byte, 2)
+	log.SetWriter("tx_write_test", writerFunc(func(p []byte) (n int, err error) {
+		logCh <- p
```

Comment:
The reason the test gets stuck is because the SetWriter persists between tests. Because of that, tests that run after this test will write to logCh, which causes it to block because no one is reading from the channel anymore.

The fix is to just clear the log writer at the end of this test.